log: Store http request/responses in a log file

vadmeste opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

`export MINIO_HTTP_TRACE=1` will log all the HTTP request/responses to stdout

## Motivation and Context
Log HTTP requests + responses

## How Has This Been Tested?
Manually

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

vadmeste committed:
log: Store http request/responses in a log file

When MINIO_TRACE_DIR is provided, create a new log file and store all
HTTP requests + responses data, body are excluded to reduce memory
consumption.

harshavardhana commented on cmd/gateway-main.go:
```diff
@@ -365,8 +365,10 @@ func gatewayMain(ctx *cli.Context, backendType gatewayBackend) {
 		setAuthHandler,
-		// Add new handlers here.
+		// Add HTTP logging handler,
+		setLoggingHandler,
```
We shouldn't even add this handler if ENV is not enabled.

vadmeste replied:
> We shouldn't even add this handler if ENV is not enabled.

If ENV is not enabled, this line doesn't have any effect actually since setLoggingHandler won't create any new middleware. But you are right, it is probably more clear to check ENV here in this place.

krishnasrinivas replied:
having it as part of the middleware will be useful in enabling logging using `mc admin` command without having to restart minio.

balamurugana commented on cmd/generic-handlers.go:
```diff
@@ -469,3 +474,67 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	return loggingHandler{
+		writer:     logFile,
+		handler:    h,
+		timeFormat: "2006-01-02 15:04:05 -0700",
```
I think `timeFormat` field is not required.  You could use `"2006-01-02 15:04:05 -0700"` format inside the handler.

vadmeste replied:
> I think timeFormat field is not required.

Why? Can you elaborate further ?

balamurugana replied:
As the time format is hard coded here, I don't think the field is needed

vadmeste replied:
> As the time format is hard coded here, I don't think the field is needed

Fixed

balamurugana commented on cmd/generic-handlers.go:
```diff
@@ -469,3 +474,67 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+// Enable HTTP requests and responses logging handler
+func setLoggingHandler(h http.Handler) http.Handler {
+	logDir := os.Getenv("MINIO_TRACE_DIR")
```
Why do we prefer log directory than log file?

vadmeste replied:
> Why do we prefer log directory than log file?

Personally, I don't prefer saving all logs in only one file, each time Minio is executed, a new file with a timestamped name will be created.

balamurugana replied:
OK. got it.

balamurugana commented on cmd/generic-handlers.go:
```diff
@@ -469,3 +474,68 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Log HTTP Response
+	respDump, err := httputil.DumpResponse(rec.Result(), false)
+	if err == nil {
+		respTime := time.Now().Format(timeFormat) + " << "
```
When parallel request/response are logged, the relation between one request and its response it not properly related here.

You would need to create a request ID kind of relation (I am OK this as `time.Now().Format(timeFormat)`) when dumping response.

krishnasrinivas reviewed (commented):
Optional ability to dump request/response body will be useful as well.

For example, if a completemultipart is failing and we need to know what all parts were provided in the request to debug further.

balamurugana commented on cmd/generic-handlers.go:
```diff
@@ -469,3 +474,68 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Create the log file
+	logFile, err := os.Create(filepath.Join(logDir, "minio-"+time.Now().Format("2006-01-02T15:04:05")+".log"))
+	if err != nil {
+		errorIf(err, "Cannot create the log file")
```
You would need to return `h` here.

balamurugana commented on cmd/generic-handlers.go:
```diff
@@ -469,3 +477,77 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+func (h loggingHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Time that will be printed in log
+	timeFormat := "2006-01-02 15:04:05 -0700"
```
This needs to be a constant

balamurugana commented on cmd/generic-handlers.go:
```diff
@@ -469,3 +477,77 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Generate short random request ID
+	randBytes := []byte("0000")
+	_, _ = io.ReadFull(rand.Reader, randBytes[:])
+	reqID := fmt.Sprintf("%X", randBytes)
```
IMO we could use `time.Now().UnixNano()` as request ID.  This is better than random number and better debugability.

balamurugana commented on cmd/generic-handlers.go:
```diff
@@ -469,3 +477,77 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		h.writer.Write([]byte(reqBanner))
+		h.writer.Write(requestDump)
+		h.writer.Write([]byte("\n"))
+		h.writeLock.Unlock()
```
`writeLock` can be removed if single `Write()` is done.  May be `bytes.Buffer` can be used.

vadmeste replied:
> writeLock can be removed if single Write() is done. May be bytes.Buffer can be used.

Are you sure @balamurugana ? we are writing in a file, I am not sure if the OS or golang synchronizes parallel writing.

balamurugana commented on cmd/routers.go:
```diff
@@ -113,6 +114,11 @@ func configureServerHandler(endpoints EndpointList) (http.Handler, error) {
 	}
 
+	// Add HTTP requests/responses logging handler when ENV is et
+	if os.Getenv("MINIO_TRACE_DIR") != "" {
```
env variables need to be handled in `handleCommonEnvVars()`

vadmeste committed:
log: Use non mem consuming http req/resp recorders

httputil recorder consumes memory when we want to log the first N bytes
of a request or response bodies. This commit adds new recorders that are
not memory consumers. The maximum is about 32k per request.

balamurugana commented on cmd/generic-handlers.go:
```diff
@@ -469,3 +477,90 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	)
+
+	h.writeLock.Lock()
+	h.writer.Write([]byte(logText))
```
As we log request/response together, till the response is returned, the request is not logged.  Similarly if response is not returned, the request is not at all logged.

vadmeste replied:
> As we log request/response together, till the response is returned, the request is not logged. Similarly if response is not returned, the request is not at all logged.

yes totally right, will fix it.

balamurugana commented on cmd/generic-handlers.go:
```diff
@@ -469,3 +477,90 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	const recordLen = 32 * 1024
+
+	// Generate short random request ID
+	reqID := fmt.Sprintf("%X", time.Now().UnixNano()%1e8)
```
I think we are decided to do `dmesg` style request id.

vadmeste replied:
> I think we are decided to do dmesg style request id.

oh yes.. but then I thought isn't it better to have complete time (full date, time) than seconds since boot time ? (sorry for not bringing this up).

balamurugana replied:
How about printing `time.UnixNano()` with a `.` after 10th digit like `1502748388.583970878`?  My worry is rolling over request id.

balamurugana reviewed (commented):
Removing -1

balamurugana reviewed (commented):
Removing -1

krishnasrinivas commented on cmd/generic-handlers.go:
```diff
@@ -469,3 +477,90 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	}
+	w.WriteHeader(rec.Code)
+	// Copy the captured response body to the ew response
+	_, wErr := rec.Body.WriteTo(w)
```
there is an issue with this implementation - clients do not receive response beyond 32K

krishnasrinivas replied:
Alternative way to implement: https://github.com/krishnasrinivas/httpdump/blob/master/main.go

vadmeste replied:
> there is an issue with this implementation - clients do not receive response beyond 32K

it does @krishnasrinivas, respBodyRecorder has the 32k limitation not rec.Body

your implementation alternative is wonderful, but we will need to include another dependency (forward), are we okay with that ?

krishnasrinivas replied:
```
+	rec := httptest.NewRecorder()
+	respBodyRecorder := &recordResponseWriter{ResponseWriter: rec, recLen: recordLen}
```

this would mean that PutObjectHandler writes to `rec` and hence all the contents will have to be stored in memory, right?

> your implementation alternative is wonderful, but we will need to include another dependency (forward), are we okay with that ?

`httpdump` is not written to be included as a dependency, I was just pointing out to an alternate implementation if httptest.NewRecorder() does not solve memory problem

vadmeste replied:
> this would mean that PutObjectHandler writes to rec and hence all the contents will have to be stored in memory, right?

@krishnasrinivas, you mean GetObjectHandler not PutObjectHandler.. yeah you are right.. I was checking putting object but it turned out httptest Response Recorder saves everything.. in contrary to request.Body saves in memory only when asked to. I'll see your alternative solution, thanks for suggesting it.

krishnasrinivas commented on cmd/generic-handlers.go:
```diff
@@ -469,3 +475,75 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	writer    io.Writer
+	writeLock *sync.Mutex
+	handler   http.Handler
+}
```
using sync.Mutex like this is easier:
```
+type loggingHandler struct {
 +	writer    io.Writer
 +	sync.Mutex
 +	handler   http.Handler
 +}
```

krishnasrinivas replied:
more info: https://twitter.com/davecheney/status/743289114980548612

using `sync.Mutex` instead of `*sync.Mutex` means we have to do pointer receiver:
func (h *loggingHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {

krishnasrinivas commented on cmd/generic-handlers.go:
```diff
@@ -469,3 +475,75 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		string(respBodyRecorder.Data()))
+	h.writeLock.Lock()
+	h.writer.Write([]byte(logText))
+	h.writeLock.Unlock()
```
you can use one lock/unlock block and inside use fmt.Fprintf() to log both req/resp (instead of fmt.Sprintf) to be more mem efficient

krishnasrinivas committed:
Log to stdout. Disable body logging for PutObject PutObjectPart GetObject.

MINIO_HTTP_TRACE=1 enables logging.

krishnasrinivas committed:
Removing testcase as capturing the stdout is tricky

krishnasrinivas committed:
Add comments and move related functions to http-trace.go

krishnasrinivas committed:
Add comments

harshavardhana commented on cmd/http-trace.go:
```diff
@@ -0,0 +1,219 @@
+/*
+ * Minio Cloud Storage, (C) 2015, 2016, 2017 Minio, Inc.
```
New file just add 2017

harshavardhana commented on cmd/lock-instrument_test.go:
```diff
@@ -367,6 +368,9 @@ func TestNsLockMapStatusBlockedToRunning(t *testing.T) {
 	}
 
+	// Initialize the locks.
+	initNSLock(false)
```
why is this needed?

vadmeste replied:
> why is this needed?

Umm, if I don't put this, TestNsLockMapStatusBlockedToRunning will fail.

This is also done in other similar unit tests. But frankly, I don't know too much low level to fully confirm why this is needed.

harshavardhana replied:
Don't know we shouldn't need to do this.. there is something else going wrong.

vadmeste committed:
Move http log tracer to a separate pkg

harshavardhana commented on cmd/gateway-main.go:
```diff
@@ -418,8 +418,8 @@ func gatewayMain(ctx *cli.Context, backendType gatewayBackend) {
 		// invalid/unsupported signatures.
 		setAuthHandler,
-		// Add new handlers here.
 
```
avoid this change?

harshavardhana commented on cmd/common-main.go:
```diff
@@ -114,4 +114,8 @@ func handleCommonEnvVars() {
 		globalIsBrowserEnabled = bool(browserFlag)
 	}
+
+	if os.Getenv("MINIO_HTTP_TRACE") != "" {
```
why not simply `globalHTTPTrace == os.Getenv("MINIO_HTTP_TRACE") != ""` ?

harshavardhana commented on cmd/http-trace.go:
```diff
@@ -0,0 +1,219 @@
+	// Internal recording buffer
+	buf bytes.Buffer
+}
+
```
@vadmeste can we move this to a generic package?

harshavardhana commented on cmd/http-trace.go:
```diff
@@ -0,0 +1,219 @@
+
+// Logs request/response headers and body.
+// Body is not logged for PutObject, PutObjectPart and GetObject.
+func httpTracelog(f http.HandlerFunc, logBody bool) http.HandlerFunc {
```
We should call this as handler since it returns a HandlerFunc .. like `httpTraceHandler` etc.

harshavardhana commented on cmd/http-trace.go:
```diff
@@ -0,0 +1,219 @@
+// Body is not logged for PutObject, PutObjectPart and GetObject.
+func httpTracelog(f http.HandlerFunc, logBody bool) http.HandlerFunc {
+	name := runtime.FuncForPC(reflect.ValueOf(f).Pointer()).Name()
+	name = strings.TrimPrefix(name, "github.com/minio/minio/cmd.")
```
wouldn't `getSource` help here? or a similar code..

harshavardhana commented on cmd/test-utils_test.go:
```diff
@@ -2392,3 +2392,71 @@ func TestToErrIsNil(t *testing.T) {
+
+// prepareTestServer - helper function that setups a single-node FS backend
+// for generic handlers tests.
+func prepareFSTestServer(mux *router.Router) (*fsTestBed, error) {
```
what is this for?

harshavardhana commented on cmd/gateway-router.go:
```diff
@@ -67,63 +67,63 @@ func registerGatewayAPIRouter(mux *router.Router, gw GatewayLayer) {
+	bucket.Methods("POST").HeadersRegexp("Content-Type", "multipart/form-data*").HandlerFunc(httpTracelogAll(api.PostPolicyBucketHandler))
 	// DeleteMultipleObjects
-	bucket.Methods("POST").HandlerFunc(api.DeleteMultipleObjectsHandler)
+	bucket.Methods("POST").HandlerFunc(httpTracelogAll(api.DeleteMultipleObjectsHandler))
```
Once we move `httpTraceLog` to its own package we can simplify these and their naming conventions.

krishnasrinivas commented on cmd/handler-utils.go:
```diff
@@ -240,3 +242,19 @@ func extractPostPolicyFormValues(form *multipart.Form) (filePart io.ReadCloser,
+		return f
+	}
+	return httpTracer.TraceReqHandlerFunc(f, false)
+}
```
you can move these functions also into the pkg?

krishnasrinivas replied:
oops, I see that it accesses globalHTTPTrace

harshavardhana commented on cmd/lock-instrument_test.go:
```diff
@@ -289,6 +290,9 @@ func TestNsLockMapStatusBlockedToRunning(t *testing.T) {
 	}
 
+	// Initialize the locks.
+	// initNSLock(false)
```
Should we not remove this if not needed?

harshavardhana commented on pkg/http-tracer/http-trace.go:
```diff
@@ -0,0 +1,202 @@
+/*
```
Keep this package to be `handlers` instead of http-trace is too high level, also please bring a test case for this package. `handlers.TraceReqHandlerFunc` , we can move few other handlers into this package as well.  Eventually..

harshavardhana commented on pkg/http-tracer/http-trace.go:
```diff
@@ -0,0 +1,202 @@
+ * limitations under the License.
+ */
+
+package cmd
```
Please fix this :cry:
